contracts/mesh-provider/src/ibc.rs

```diff
@@ -111,26 +118,53 @@ pub fn ibc_packet_receive(
 
     let msg: ConsumerMsg = from_slice(&msg.packet.data)?;
     match msg {
-        ConsumerMsg::Rewards {} => receive_rewards(deps),
+        ConsumerMsg::Rewards {
+            validator,
+            total_funds,
+        } => receive_rewards(deps, env, validator, total_funds),
         ConsumerMsg::UpdateValidators { added, removed } => {
-            receive_update_validators(deps, added, removed)
+            receive_update_validators(deps, env, added, removed)
         }
     }
 }
 
-pub fn receive_rewards(_deps: DepsMut) -> Result<IbcReceiveResponse, ContractError> {
-    // TODO
-    unimplemented!();
+pub fn receive_rewards(
+    deps: DepsMut,
+    env: Env,
+    validator: String,
+    total_funds: Coin,
+) -> Result<IbcReceiveResponse, ContractError> {
+    // Update the rewards of this validator
+    // This will fail if we didn't add the validator before it, we cannot init the validator and calculate rewards in the same msg. (same block)
+    let res = VALIDATORS.update::<_, ContractError>(deps.storage, &validator, |val| {
+        let mut val = match val {
+            Some(val) => val,
+            None => return Err(ContractError::ValidatorRewardsCalculationWrong {}),
+        };
+
+        val.rewards.calc_rewards(total_funds.amount, val.shares_to_tokens(val.stake), env.block.height)?;
+
+        Ok(val)
+    });
+
+    // TODO: if calculation failed, we want to handle it as leftover funds? or send funds back to consumer and handle it there?
+    let ack = match res {
+        Ok(_) => to_binary(&StdAck::success(&RewardsResponse {}))?,
+        Err(err) => to_binary(&StdAck::Error(err.to_string()))?,
```

Comment:
Note (as of ibcv3, which we can count on here), this is now handled in the wasmd framework. 
If you return error here, it is turned into StdAck::Err.

You can just do something like:
```
// actually put this at the end of the ...update statement
let res = res?;
let ack = to_binary(&StdAck::success(&RewardsResponse {}))?;
```

Reply:
Nice, didn't know it was handled already.